pkg/agent/wireguard/client_test.go

```diff
@@ -0,0 +1,415 @@
+//go:build linux
+
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wireguard
+
+import (
+	"net"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+
+	"antrea.io/antrea/pkg/agent/config"
+)
+
+type fakeWireGuardClient struct {
+	peers map[wgtypes.Key]wgtypes.Peer
+}
+
+func (f *fakeWireGuardClient) Close() error {
+	return nil
+}
+
+func (f *fakeWireGuardClient) Devices() ([]*wgtypes.Device, error) {
+	return nil, nil
+}
+
+func (f *fakeWireGuardClient) Device(name string) (*wgtypes.Device, error) {
+	var res []wgtypes.Peer
+	for _, p := range f.peers {
+		res = append(res, p)
+	}
+	return &wgtypes.Device{
+		Peers: res,
+	}, nil
+}
+
+func (f *fakeWireGuardClient) ConfigureDevice(name string, cfg wgtypes.Config) error {
+	for _, c := range cfg.Peers {
+		if c.Remove {
+			delete(f.peers, c.PublicKey)
+		} else {
+			peer := wgtypes.Peer{
+				PublicKey: c.PublicKey,
+				Endpoint:  c.Endpoint,
+			}
+			if c.Endpoint != nil {
+				ep := &net.UDPAddr{}
+				ip := c.Endpoint.IP
+				ep.IP = append(ip[:0:0], ip...)
+				ep.Port = c.Endpoint.Port
+				ep.Zone = c.Endpoint.Zone
+				peer.Endpoint = ep
+			}
+			for _, n := range c.AllowedIPs {
+				peer.AllowedIPs = append(peer.AllowedIPs, net.IPNet{
+					IP:   append(n.IP[:0:0], n.IP...),
+					Mask: append(n.Mask[:0:0], n.Mask...),
+				})
+			}
+			f.peers[c.PublicKey] = peer
```

Comment:
Any reason the code is not simple as below?
```
f.peers[c.PublicKey] = wgtypes.Peer{
	PublicKey: c.PublicKey,
	Endpoint:  c.Endpoint,
	AllowedIPs: c.AllowedIPs,
}
```

Reply:
I think the input IP/IPNet won't change as it's not common to update an IP or IPNet in-place. Otherwise all code that cache IP/IPNet would need to make a copy.

Reply:
I just want to make sure those values are deep copied.

Reply:
OK. Updated accordingly.